File: 2255-minimum-swaps-to-group-all-1s-together-ii/2255-minimum-swaps-to-group-all-1s-together-ii.java
class Solution {
    public int minSwaps(int[] nums) {
        int total1=0; // total number of 1 in nums
        int count=0; // total number of 1 in window of size total1
        int n=nums.length;
        for(int i=0;i<n;i++){
            total1+=nums[i];
        }
        for(int i=0;i<total1;i++){
            count+=nums[i];
        }
        int maxCount=count; //total number of 1 in any window of size total1
        for(int i=1;i<n;i++){
            count-=nums[i-1];
            count+=nums[(i+total1-1)%n];
            maxCount=Math.max(maxCount,count);
        }
        return total1-maxCount;
    }
}

File: 2255-minimum-swaps-to-group-all-1s-together-ii/2255-minimum-swaps-to-group-all-1s-together-ii.cpp
class Solution {
public:
    int minSwaps(vector<int>& nums) {
        int total1 = 0; // total number of 1 in vector
        int count = 0;  // total number of 1 in window of size total1
        int max1 = 0;   // maximum number of 1 in any window of size total1
        int n = nums.size();
        for (int i = 0; i < nums.size(); i++) {
            total1 += nums[i];
        }
        for (int i = 0; i < total1; i++){
            count += nums[i];
        }
        max1 = count;
        for (int i = 1; i < n; i++) {
            count += nums[(i + total1 - 1) % n];
            count -= nums[i - 1];
            max1 = max(max1, count);
        }
        return total1 - max1;
    }
};